### DISCUSSION TITLE
Schema validation in tests

### DISCUSSION THREAD

**Discussion by AmusedArg** [L2jLiga/fastify-decorators#800] [Q&A] [answered]
Hi, I'm having trouble while testing my app, if I start my application it loads the schemas correctly and they're working. 
But when I run the tests it throws the following error: _**FastifyError: Failed building the validation schema for GET: /api/lorem/ipsum, due to error can't resolve reference default.authorization.header# from id #**_

The test fails in the before all when creating the configureControllerTest like this
```javascript
describe('Controller: CalculatorController', () => {
  let instance: FastifyInstanceWithController<CalculatorController>;
  const myService = {someFunction: jest.fn()};
  beforeAll(async () => {
    instance = await configureControllerTest({
      controller: CalculatorController,
      mocks: [
        {
          provide: MyService,
          useValue: myService,
        },
      ],
    });
  });
});
```
And the controller definition is like this this:
```javascript
@Controller()
export default class CalculatorController {
    @GET('/api/lorem/ipsum', {
      schema: {
        headers: {$ref: 'default.authorization.header#'}
      },
    })
    init() { /*some code*/ }
}
```

Thanks!

**Comment by L2jLiga** [comment, marked as answer]
Hi there!

I think you can workaround this issue in tests by using `plugins` option. Something like this I guess
```typescript
describe('Controller: CalculatorController', () => {
  let instance: FastifyInstanceWithController<CalculatorController>;
  const myService = {someFunction: jest.fn()};
  beforeAll(async () => {
    instance = await configureControllerTest({
      controller: CalculatorController,
      mocks: [
        {
          provide: MyService,
          useValue: myService,
        },
      ],
      plugins: [
        (instance) => instance.addSchema(authorizationHeaderSchema)
      ]
    });
  });
});
```


If this schema comming from another plugin then you can replace this "self written plugin" with actual one

**Reply by AmusedArg** [reply]
Hi!
It's still throwing the same error :(

```typescript
   const defaultAuthorizationHeaderSchema = {
      $id: 'default.authorization.header',
      type: 'object',
      required: ['authorization'],
    };
    instance = await configureControllerTest({
      controller: CalculatorController,
      mocks: [
        {
          provide: ApiGatewayService,
          useValue: apiGatewayService,
        },
      ],
      plugins: [
        (instance: any) => instance.addSchema(defaultAuthorizationHeaderSchema),
      ],
    });
```

**Reply by L2jLiga** [reply]
Yeah, my bad, schemas are encapsulated within  plugin and won't be available out of it, hence there's no option to provide shared schemas atm.

Possible short-term solution is to take this function and reimplement it in your tests or patch it somehow (f.e with [patch-package](https://www.npmjs.com/package/patch-package)) to make it possible to add shared schemas
https://github.com/L2jLiga/fastify-decorators/blob/v3/lib/testing/configure-controller-test.ts#L35-L60

Long-term it would be nice to have option to configure such things as shared schemas, Ajv instance and so on

**Reply by AmusedArg** [reply]
I will try that. Thanks anyway!!

**Reply by L2jLiga** [reply]
I've created issue for that, may be solution will be to add optional field "fastifyInstance" to configuration in order to provide own instance with Ajv/schemas and so on configured
https://github.com/L2jLiga/fastify-decorators/issues/801

**Reply by L2jLiga** [reply]
I've released [v3.14.0](https://github.com/L2jLiga/fastify-decorators/releases/tag/v3.14.0)

In new version you can provide `instane` option which is `FastifyInstance` so you can easily preconfigure your instance for tests with all required schemas and so on..

**Reply by AmusedArg** [reply]
Nice!

But I'm having another issue now, when passing my instance to the configureControllerTest() I'm getting the following error:

Method 'GET' already declared for route '/api/:/suggestions' with constraints '{}'

**Reply by L2jLiga** [reply]
I guess you're passing instance from your main file, if so this will not work because it's already has registered routes, in order to make it work you have to create new instance within your test, preferably in same book and setup your instance with everything needed.

**Reply by L2jLiga** [reply]
Anyway it might be good idea for improving testing experience to provide option to mock services in already set up instance but I have no idea how to make it with current library structure and behavior

**Reply by AmusedArg** [reply]
Yup I  was using that instance.

I will try that then. Thank you very much!!
